Feature/telescope programming tutorial

tuanchien opened the pull request:
There are 4 main additions in this PR.

1. Telescope programming tutorial.
2. Command line tool to generate a fresh telescope template. See `observatory generate telescope --help` for more details.
3. BigQuery Schema JSON to CSV converter.
4. Changing the hard coded schemas in the documentation to use tables created by the CSV files instead.

Would appreciate comments on the programming tutorial. Aniek mentioned she was also working on a tutorial that explains the difference between the telescope types in more detail, so currently the part that describes the different architectures is a bit more bare bones.

jdddog reviewed (commented):
Hey Tuan, it looks really good.

There are a couple of things that I suggest to make the schema documentation easier to follow:
1) Flatten the layout hierarchy as much as possible.
On the Datasets page, we could flatten the layout to the following:
```bash
Crossref Metadata
Crossref Events
Fundref
GeoNames
GRID
Microsoft Academic Graph
Unpaywall
Web of Science
Scopus
```

Then, on each specific dataset page, it would be good to directly show the latest schema, rather than having that in a `schemas` sub page. For instance (if it was written in markdown):
```bash
## GeoNames
“The GeoNames geographical database covers all countries.” “Users may manually edit, correct and add new names.” - source GeoNames

“The GeoNames geographical database covers all countries.” It contains “over 25 million geographical names and consists of over 11 million unique features whereof 4.8 million populated places and 13 million alternate names. All features are categorized into one out of nine feature classes and further subcategorized into one out of 645 feature codes. GeoNames is integrating geographical data such as names of places in various languages, elevation, population and others from various sources. - source: GeoNames and data details

| name      | type | mode      | description |
| ----------- | ----------- | ----------- | ----------- |
| geonameid      | INTEGER       | NULLABLE       | integer id of record in geonames database 
```

2) Only show the latest version of the schema.
It would make the documentation easier to follow by only showing the latest version of each schema. I don't think people really need to see all of the different versions. If you want to find out what is different between schemas you would use a file diffing tool anyway.

rhosking commented on docs/tutorials/telescope_prog_intro.md:
```diff
@@ -0,0 +1,254 @@
+Each workflow in the Academic Observatory consists of set of tasks.  The tasks are organised as nodes on a directed acyclic graph (DAG) which describe the order of execution, and dependencies. Telescope tasks are Python functions with a specific signature, that are turned into Airflow PythonOperator objects.  The exception is the sensor tasks, which are Airflow Sensor objects, e.g., ExternalTaskS ...
+
+The telescope class groups tasks into:
+- sensor tasks,
```
I'm not sure this is displaying correctly, did you mean for this to be a bullet point list?

rhosking reviewed (commented):
this looks nice Tuan, good work!

For the generate command, what are your thoughts on also creating files for testing and documentation with some basic structure by default?

tuanchien commented:
@jdddog I've flattened the toc depth to 1 in the dataset index files.  Removed old schemas and a few of the subsections, but am still keeping 1 heading with the schema date. For your geonames example, the index.rst has a brief description and is then linking a md for another description.  It looks like this was done for when there are several datasets for the same provider.  Do you want cases like geonames to just use the index file so that everything shows up on one page?

@rhosking I changed the list formatting.  You're right, it didn't look correct.  Templates for a test and documentation file seem reasonable. I can add that next.

tuanchien commented:
Added doc/test template generation.

Not sure of a good way to refer to the docs/test directory so it's currently a hardcoded relative path.

jdddog commented:
> @jdddog I've flattened the toc depth to 1 in the dataset index files. Removed old schemas and a few of the subsections, but am still keeping 1 heading with the schema date. For your geonames example, the index.rst has a brief description and is then linking a md for another description. It looks like this was done for when there are several datasets for the same provider. Do you want cases like geonames to just use the index file so that everything shows up on one page?
> 
Yeah 1 heading with the schema date is good. 

It would simplify the presentation to ignore the idea of a 'provider' and just present each unique dataset (e.g. Crossref Metada, Crossref Events, Fundref, GRID etc) as an item in the top level menu and show the schema information as soon as you click on the item, otherwise there are too many levels to navigate through to get to the schema information.

tuanchien commented:
Providers have been removed.  Now in the root dataset page there is just a list of individual datasets.

CSV generation now also generates a schema_latest.csv for each schema.  The documentation links now point to the latest.csv versions rather than a specific version so that the documentation automatically updates after new schema versions are added.

tuanchien commented:
- Merged datasets into telescopes documentation.
- Removed the datasets directory. It might be less confusing to omit it rather than have an empty datasets page, until we add content on the COKI datasets.
- Add latest schemas to the telescopes that have one, but weren't in the datasets documentation.

jdddog commented on docs/telescopes/scopus.md:
```diff
@@ -1,6 +1,13 @@
 # SCOPUS
```
```suggestion
# Scopus
```

jdddog commented on docs/tutorials/telescope_prog_intro.md:
```diff
@@ -0,0 +1,247 @@
+
+The interface is defined in `AbstractRelease`. There is also a `Release` class you can use which implements all of those functions if you want to just let the API generate all those folders and file lists based on the `dag_id` and `release_id`.  See the API documentation for the `Release` class for more information.  Your telescope file should contain a **release** class if it's required for your ...
+
+#### Telescope class
```
```suggestion
## Telescope class
```

jdddog commented on docs/tutorials/telescope_prog_intro.md:
```diff
@@ -0,0 +1,247 @@
+
+The interface specification is defined in `AbstractTelescope`.  There is a `Telescope` class you can use with an implementation of that interface. Your telescope file should contain a telescope class.  See the API documentation for more details. The telesecope class is responsible for adding tasks to the DAG, constructing a DAG for Airflow to use, and onstructing release objects to pass to each t ...
+
+#### Example telescope file
```
```suggestion
### Example telescope file
```

jdddog commented on docs/tutorials/telescope_prog_intro.md:
```diff
@@ -0,0 +1,247 @@
+
+If you start the observatory platform with these changes, you will see a new DAG in Airflow called **my_dag_id** with the DAG structure `task1 -> task2 -> cleanup`.  When you run the DAG, you should see the logging messages in the log of each task.
+
+#### BigQuery schemas
```
```suggestion
### BigQuery schemas
```

jdddog commented on docs/tutorials/telescope_prog_intro.md:
```diff
@@ -0,0 +1,247 @@
+
+The BigQuery table loading utility functions in the Academic Observatory platform API will try to find the correct schema to use for loading table data, based on release date information.
+
+#### Specialised telescopes
```
```suggestion
### Specialised telescopes
```

jdddog commented on docs/tutorials/telescope_prog_intro.md:
```diff
@@ -0,0 +1,247 @@
+
+If you wish to design your own telescope templates, you can choose to derive the `Telescope` class if it suits your needs, or implement the `AbstractTelescope` interface yourself.
+
+### Generating a telescope template
```
```suggestion
## Generating a telescope template
```

jdddog commented on docs/tutorials/telescope_prog_intro.md:
````diff
@@ -0,0 +1,247 @@
+```
+creates the files `observatory-dags/observatory/dags/dags/mynewtelescope.py` and `observatory-dags/observatory/dags/telescopes/mynewtelescope.py`
+
+### Testing
````
```suggestion
## Testing
```

jdddog commented on docs/tutorials/telescope_prog_intro.md:
```diff
@@ -0,0 +1,247 @@
+
+You can test telescopes by running them in Airflow.  The Academic Observatory has additional infrastructure to facilitate unit testing of telescopes without needing to run the telescopes in Airflow.
+
+### Documentation
```
```suggestion
## Documentation
```

jdddog commented on docs/tutorials/telescope_prog_intro.md:
````diff
@@ -0,0 +1,247 @@
+```
+This will output html documentation in the `docs/_build/html` directory.
+
+#### Including schemas in documentation
````
```suggestion
### Including schemas in documentation
```

jdddog commented on docs/tutorials/telescope_prog_intro.md:
````diff
@@ -0,0 +1,247 @@
+```
+The `..` follows the parent directory, and we need to do this twice to reach `docs` from `docs/datasets/mydataset`.
+
+### Style
````
```suggestion
## Style
```

jdddog commented on docs/tutorials/telescope_prog_intro.md:
```diff
@@ -0,0 +1,247 @@
+
+We recommend you use those format tools as part of your coding workflow.
+
+##### Type hinting
```
```suggestion
### Type hinting
```

jdddog commented on docs/tutorials/telescope_prog_intro.md:
```diff
@@ -0,0 +1,247 @@
+
+You should provide type hints for all of the function arguments you use, and for return types.  Because Python is a weakly typed language, it can be confusing to those unacquainted with the codebase what type of objects are being manipulated in a particular function.  Type hints help reduce this ambiguity.
+
+##### Docstring
```
```suggestion
### Docstring
```

jdddog commented on docs/tutorials/telescope_prog_intro.md:
```diff
@@ -0,0 +1,247 @@
+
+### Testing
+
+You can test telescopes by running them in Airflow.  The Academic Observatory has additional infrastructure to facilitate unit testing of telescopes without needing to run the telescopes in Airflow.
```
I think either remove this section, or provide a link to the full Testing section.

jdddog commented on docs/tutorials/telescope_prog_testing.md:
```diff
@@ -0,0 +1,126 @@
+
+To take advantage of the framework, instead of inheriting `unittest.TestCase` in test classes, each test class inherits `ObservatoryTestCase`.
+
+## Testing DAGs
```
Maybe split it out into three main subheadings under `Testing`:

* Testing DAG structure
* Testing DAG loading
* Functional testing

jdddog reviewed (commented):
Hey Tuan, it looks good, I've left some small comments. I think we can merge it after addressing those and rebasing.

tuanchien commented:
Added new commit that hopefully addresses Jamie's comments
